File: legacy/src/scenes/dungeon/maze/nearest-open-tile.ts
/**
 * Find open floor near a tile — the "put this thing SOMEWHERE sensible" helper
 * every spawner reaches for.
 *
 * Moved out of core.ts: ten call sites across level building, corpse piles,
 * merchants and pin crews all want it, and none of them are core's business.
 */
import { isWalkable, type Grid, type TilePos } from "./generator";

/**
 * The `n`-th walkable tile found scanning outward in ring shells from (ci, cj).
 *
 * NOTE the semantics: `n` is an ORDINAL, not a distance. Asking for n = 6 does
 * NOT get you a tile 6 tiles out — it gets the 6th walkable tile found, which
 * in an open area is still inside the r = 1 ring. Pass `minRing` when you
 * actually mean "no closer than this".
 */
export function nearestOpenTile(g: Grid, ci: number, cj: number, n: number, minRing = 1): TilePos | null {
  const found: TilePos[] = [];
  for (let r = Math.max(1, minRing); r <= Math.max(6, minRing + 5) && found.length < n; r++) {
    for (let dj = -r; dj <= r; dj++) {
      for (let di = -r; di <= r; di++) {
        if (Math.max(Math.abs(di), Math.abs(dj)) !== r) continue; // ring shell only
        const i = ci + di;
        const j = cj + dj;
        if (isWalkable(g, i, j)) found.push({ i, j });
        if (found.length >= n) break;
      }
    }
  }
  return found[n - 1] ?? found[found.length - 1] ?? null;
}

File: crates/pk-core/src/maze/nearest_open_tile.rs
//! Outward Ring Shell Walkable Tile Scanner — Spatial query for nearest open ground.
//!
//! PORTS: `maze/nearest-open-tile.ts`

use crate::grid::{is_walkable, Grid};
use crate::maze::track_launch::TilePos;

/// The `n`-th walkable tile found scanning outward in ring shells from (ci, cj).
///
/// `n` is an ordinal index (1-based), not a distance.
pub fn nearest_open_tile(g: &Grid, ci: i32, cj: i32, n: usize, min_ring: usize) -> Option<TilePos> {
    if n == 0 {
        return None;
    }
    let mut found = Vec::new();
    let r_start = (min_ring.max(1)) as i32;
    let r_end = (6_i32).max(min_ring as i32 + 5);

    for r in r_start..=r_end {
        if found.len() >= n {
            break;
        }
        for dj in -r..=r {
            for di in -r..=r {
                if di.abs().max(dj.abs()) != r {
                    continue;
                }
                let i = ci + di;
                let j = cj + dj;
                if is_walkable(g, i, j) {
                    found.push(TilePos { i, j });
                }
                if found.len() >= n {
                    break;
                }
            }
            if found.len() >= n {
                break;
            }
        }
    }

    if n <= found.len() {
        Some(found[n - 1])
    } else {
        found.last().copied()
    }
}
